monkeytypes/credentials/credentials.py

```diff
@@ -1,15 +1,65 @@
 from __future__ import annotations
 
-from typing import Optional, Union
+from typing import Annotated, Optional, Union
 
-from pydantic import field_serializer
+from pydantic import Discriminator, Tag, field_serializer
 
-from .. import InfectionMonkeyBaseModel
+from .. import INVALID_UNION_MEMBER_ERROR, InfectionMonkeyBaseModel
 from . import EmailAddress, LMHash, NTHash, Password, SSHKeypair, Username
 from .encoding import get_plaintext
 
-Identity = Union[Username, EmailAddress]
-Secret = Union[Password, LMHash, NTHash, SSHKeypair]
+
+def get_discriminator_value_identity(v) -> Optional[str]:
+    keys = list(v.keys())
+
+    if len(keys) == 0:
+        return None
+
+    if keys[0] in ["username", "email_address"]:
+        return keys[0]
+
+    return None
+
+
+def get_discriminator_value_secret(v) -> Optional[str]:
+    keys = list(v.keys())
+
+    if len(keys) == 0:
+        return None
+
+    if "private_key" in keys:
+        return "ssh_keypair"
+
+    if keys[0] in ["password", "lm_hash", "nt_hash"]:
+        return keys[0]
+
+    return None
+
+
+Identity = Annotated[
+    Union[Annotated[Username, Tag("username")], Annotated[EmailAddress, Tag("email_address")]],
+    Discriminator(
+        get_discriminator_value_identity,
+        custom_error_type=INVALID_UNION_MEMBER_ERROR,
```

Comment:
Can we use pydantic's `union_tag_not_found` instead of having our own custom one?

Reply:
NVM